elrond-wasm/src/storage/mappers/token_attributes_mapper.rs

```diff
@@ -0,0 +1,239 @@
+use super::StorageMapper;
+use crate::api::{ErrorApi, ManagedTypeApi, StorageReadApi, StorageWriteApi};
+use crate::storage::{storage_get, storage_set};
+use crate::types::BoxedBytes;
+use crate::types::TokenIdentifier;
+use alloc::vec::Vec;
+use elrond_codec::NestedEncode;
+
+const MAPPING_SUFFIX: &[u8] = b".mapping";
+const COUNTER_SUFFIX: &[u8] = b".counter";
+
+const VALUE_ALREADY_SET_ERROR_MESSAGE: &[u8] =
+    b"A value was already set for this token ID and token nonce";
+
+const UNKNOWN_TOKEN_ID_ERROR_MESSAGE: &[u8] =
+    b"Unknown token id. No attributes were set for this token ID";
+
+const VALUE_NOT_PREVIOUSLY_SET_ERROR_MESSAGE: &[u8] =
+    b"A value was not previously set fot this token ID and token nonce";
+
+const COUNTER_OVERFLOW_ERROR_MESSAGE: &[u8] =
+    b"Counter overflow. This module can hold evidence for maximum (u8::MAX - 1) different token IDs";
+
+pub struct TokenAttributesMapper<SA>
+where
+    SA: StorageReadApi + StorageWriteApi + ManagedTypeApi + ErrorApi + Clone + 'static,
+{
+    api: SA,
+    main_key: BoxedBytes,
+}
+
+impl<SA> StorageMapper<SA> for TokenAttributesMapper<SA>
+where
+    SA: StorageReadApi + StorageWriteApi + ManagedTypeApi + ErrorApi + Clone + 'static,
+{
+    fn new(api: SA, main_key: BoxedBytes) -> Self {
+        TokenAttributesMapper { api, main_key }
+    }
+}
+
+impl<SA> TokenAttributesMapper<SA>
+where
+    SA: StorageReadApi + StorageWriteApi + ManagedTypeApi + ErrorApi + Clone + 'static,
+{
+    pub fn set<T: elrond_codec::TopEncode + elrond_codec::TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+        attributes: &T,
+    ) {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+
+        let mapping = if has_mapping {
+            self.get_mapping_value(token_id)
+        } else {
+            let mut counter = self.get_counter_value();
+            if counter == u8::MAX {
+                self.api.signal_error(COUNTER_OVERFLOW_ERROR_MESSAGE);
+            }
+
+            counter += 1;
+            self.set_mapping_value(token_id, counter);
+            self.set_counter_value(counter);
+            counter
+        };
+
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
```

Comment:
Since this seems to be repeated in other functions as well, it would be best to have a separate function the "has_value" check instead of manually checking for is_empty.